sql-plugin/src/main/scala/com/nvidia/spark/rapids/SpillFramework.scala

```diff
@@ -0,0 +1,1606 @@
+    storesInternal = new SpillableStores {
+      override var deviceStore: SpillableDeviceStore = new SpillableDeviceStore
+      override var hostStore: SpillableHostStore = new SpillableHostStore(hostSpillStorageSize)
+      override var diskStore: DiskHandleStore = new DiskHandleStore(conf)
+    }
+    val hostSpillStorageSizeStr = hostSpillStorageSize.map(sz => s"$sz B").getOrElse("unlimited")
+    logInfo(s"Initialized SpillFramework. Host spill store max size is: $hostSpillStorageSizeStr.")
+  }
+
+  def shutdown(): Unit = {
+    if (hostSpillBounceBuffer != null) {
+      hostSpillBounceBuffer.close()
+      hostSpillBounceBuffer = null
+    }
+    if (chunkedPackBounceBuffer != null) {
+      chunkedPackBounceBuffer.close()
+      chunkedPackBounceBuffer = null
+    }
+    if (storesInternal != null) {
+      storesInternal.close()
+      storesInternal = null
+    }
+  }
+
+  def withHostSpillBounceBuffer[T](body: HostMemoryBuffer => T): T =
+    hostSpillBounceBuffer.synchronized {
+      body(hostSpillBounceBuffer)
+    }
+
+  def withChunkedPackBounceBuffer[T](body: DeviceMemoryBuffer => T): T =
+    chunkedPackBounceBuffer.synchronized {
+      body(chunkedPackBounceBuffer)
+    }
+
+  def removeFromDeviceStore(value: SpillableHandle): Unit = {
+    // if the stores have already shut down, we don't want to create them here
+    val deviceStore = synchronized {
+      Option(storesInternal).map(_.deviceStore)
+    }
+    deviceStore.foreach(_.remove(value))
+  }
+
+  def removeFromHostStore(value: SpillableHandle): Unit = {
+    // if the stores have already shut down, we don't want to create them here
+    val hostStore = synchronized {
+      Option(storesInternal).map(_.hostStore)
+    }
+    hostStore.foreach(_.remove(value))
+  }
+
+  def removeFromDiskStore(value: DiskHandle): Unit = {
+    // if the stores have already shut down, we don't want to create them here
+    val maybeStores = synchronized {
+      Option(storesInternal).map(_.diskStore)
+    }
+    maybeStores.foreach(_.remove(value))
+  }
+}
+
+/**
+ * ChunkedPacker is an Iterator-like class that uses a cudf::chunked_pack to copy a cuDF `Table`
+ * to a target buffer in chunks. It implements a next method that takes a DeviceMemoryBuffer
+ * as an argument to be used for the copy.
+ *
+ * Each chunk is sized at most `bounceBuffer.getLength`, and the caller should cudaMemcpy
+ * bytes from `bounceBuffer` to a target buffer after each call to `next()`.
+ *
+ * @note `ChunkedPacker` must be closed by the caller as it has GPU and host resources
+ *       associated with it.
+ *
+ * @param table cuDF Table to chunk_pack
+ * @param bounceBufferSize size of GPU memory to be used for packing. The buffer will be
+ *                         obtained during the iterator-like 'next(DeviceMemoryBuffer)'
+ */
+class ChunkedPacker(table: Table,
+                    bounceBufferSize: Long)
```

Comment:
It might be cleaner if we can make ChunkedPacker a true iterator.  It feels a little odd that it needs to be told about the size and then it's the callers responsibility to pass the exact same size buffers later to the next method, especially when we have methods where they're handed a ChunkedPacker instance.  How are they supposed to know what the correct size is for that instance?

Instead we could create a bounce buffer pool interface that looks something like this:
```scala
trait DeviceBounceBufferPool {
  // size of every buffer in the pool
  val bufferSize: Long

  // Returns the next available buffer in the pool,
  // blocking until one is available if necessary.
  // Closing the resulting buffer will release it back to the pool.
  def nextBuffer(): DeviceMemoryBuffer
}
```
Then ChunkedPacker could be constructed with a device buffer pool (potentially just a single buffer pool in the degenerate case).  This now allows it to just be an iterator.  next() will return a device buffer containing the next chunk of contents from the pool, and when the caller closes the buffer, it will be released back to the pool for reuse.

Reply:
Here's the change as suggested. I did wrap `DeviceMemoryBuffer` with a case class to support acquire/unacquire, but I can use the event handler stuff we added. Please take a look at https://github.com/NVIDIA/spark-rapids/pull/11747/commits/513d084edeba807ca4aca82d94fcd7a7c70e22db.